tutorials/migrating_ardupilot_plugin.md

```diff
@@ -0,0 +1,926 @@
+  ignition::gazebo::EntityComponentManager &_ecm,
+  ignition::gazebo::EventManager &_eventMgr);
+```
+
+Similarly, the old code provides `OnUpdate()`, which is called once per time step while simulation is running:
+
+```cpp
+// OLD
+void OnUpdate();
+```
+
+In the new code, this method is replaced by two methods, `PreUpdate()` and
+`PostUpdate()`:
+
+
+```cpp
+// NEW
+void PreUpdate(const ignition::gazebo::UpdateInfo &_info,
+  ignition::gazebo::EntityComponentManager &_ecm);
+
+void PostUpdate(const ignition::gazebo::UpdateInfo &_info,
+  const ignition::gazebo::EntityComponentManager &_ecm);
+```
+
+As the names suggest, the former is called before each time step, while the
+latter is called after each time step. Note the subtle difference in signature:
+`PreUpdate()` takes a non-`const` reference to the `EntityComponentManager`,
+while `PostUpdate()` takes a `const` reference to it. We'll make any changes to
+the state of simulation (e.g., setting torques on joints) in `PreUpdate()` and
+we'll read out results from simulation (e.g., getting the pose of a link) in
+`PostUpdate()`.
+
+The remaining changes in the header are just bookkeeping, to allow us to have
+access to the right objects with the right types in other class methods. These three helpers:
+
+```cpp
+// OLD
+void ApplyMotorForces(const double _dt);
+void SendState();
+bool InitArduPilotSockets(sdf::ElementPtr _sdf);
+```
+
+become:
+
+```cpp
+// NEW
+void ApplyMotorForces(const double _dt,
+  ignition::gazebo::EntityComponentManager &_ecm);
+void SendState(double _simTime,
+  const ignition::gazebo::EntityComponentManager &_ecm);
+bool InitArduPilotSockets(const std::shared_ptr<const sdf::Element> &_sdf);
+```
+
+## Plugin source file (ArduPilotPlugin.cc)
+
+### Headers
+
+The old code includes these Gazebo-related headers:
+
+```cpp
+// OLD
+#include <sdf/sdf.hh>
+#include <ignition/math/Filter.hh>
+#include <gazebo/common/Assert.hh>
+#include <gazebo/common/Plugin.hh>
+#include <gazebo/msgs/msgs.hh>
+#include <gazebo/sensors/sensors.hh>
+#include <gazebo/transport/transport.hh>
+```
+
+Like we did in `ArduPilotPlugin.hh`, we'll keep `<sdf/sdf.hh>`. The others are
+replaced with Ignition equivalents, and where possible we narrow the inclusion
+to exactly what we need. We start by enumerating those *components* (part of the
+ECS pattern used by Ignition) that we're using:
+
+```cpp
+// NEW
+#include <ignition/gazebo/components/AngularVelocity.hh>
+#include <ignition/gazebo/components/Imu.hh>
+#include <ignition/gazebo/components/JointForceCmd.hh>
+#include <ignition/gazebo/components/JointPosition.hh>
+#include <ignition/gazebo/components/JointVelocity.hh>
+#include <ignition/gazebo/components/JointVelocityCmd.hh>
+#include <ignition/gazebo/components/LinearVelocity.hh>
+#include <ignition/gazebo/components/Name.hh>
+#include <ignition/gazebo/components/Pose.hh>
+```
+
+Then we include the parts of `ign-gazebo` itself that we're using:
```

Comment:
This is a very helpful contribution @gerkey, thanks! It will definitely be enlightening to many developers that will hopefully migrate their plugins to Ignition Gazebo. When I first started learning the new architecture, what puzzled me most was the structure of the ECM graph (what becomes an entity and what a component?). The turning point was when I found and read the `SdfEntityCreator` class, it is the most representative example of how the ECM is populated.

This is my 2 cents:

```suggestion

A great resource to understand the logic under-the-hood of the ECM is the `SdfEntityCreator` class ([header](https://github.com/ignitionrobotics/ign-gazebo/blob/master/include/ignition/gazebo/SdfEntityCreator.hh), [source](https://github.com/ignitionrobotics/ign-gazebo/blob/master/src/SdfEntityCreator.cc)). In fact, this class is responsible to map the content of a SDF file to the entities and components that form the graph handled by the ECM. If you wonder what are the components that you can access by default from the plugin, this class is the best entry point.

Then we include the parts of `ign-gazebo` itself that we're using:
```

Reply:
Thanks for the contribution, @diegoferigo!